internal/controller/controllers/agent_controller.go

```diff
@@ -188,7 +188,32 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 	}
 
 	if agent.Spec.ClusterDeploymentName == nil && h.ClusterID != nil {
-		log.Debugf("ClusterDeploymentName is unset in Agent %s. unbind", agent.Name)
+		log.Debugf("ClusterDeploymentName is unset in Agent %s.", agent.Name)
+		installingStatuses := []string{
+			models.HostStatusPreparingForInstallation,
+			models.HostStatusPreparingFailed,
+			models.HostStatusPreparingSuccessful,
+			models.HostStatusInstalling,
+			models.HostStatusInstallingInProgress,
+			models.HostStatusInstallingPendingUserAction,
+		}
+		if funk.ContainsString(installingStatuses, *h.Status) {
+			if swag.StringValue(h.Kind) != models.HostKindAddToExistingClusterHost {
+				// Only day-2 hosts are allowed to be cancelled during installation
+				errMsg := fmt.Errorf("cancelling non-day-2 host [%s] in the middle of installation is not allowed", *h.ID)
+				log.Error(errMsg)
+				return r.updateStatus(ctx, log, agent, origAgent, &h.Host, h.ClusterID, errMsg, false)
+			}
+			// Host is in the middle of installing so cancel the installation first to allow moving to unbind
+			newHost, cancelErr := r.Installer.CancelDay2HostInstallationInternal(ctx, h)
+			if cancelErr != nil {
+				log.WithError(cancelErr).Errorf("failed cancelling installation for host %s", *h.ID)
+				return r.updateStatus(ctx, log, agent, origAgent, &h.Host, h.ClusterID, cancelErr, true)
+			}
+			h = newHost
+			log.Infof("Installation successfully cancelled for host %s", *h.ID)
+		}
+		log.Infof("Unbinding host %s from cluster %s", *h.ID, *h.ClusterID)
 		return r.unbindHost(ctx, log, agent, origAgent, h)
```

Comment:
The way this runs cancel and then immediately runs unbind has me wondering if this is worth being separate.

Nothing is going to have a chance to react to the cancelling state, right? Why not just allow unbind from the installing states directly?

Reply:
If we're not expecting any benefit out of the cancelling state (which it seems like we're not given the speed with which we move through it) then this probably makes sense.

The only issue I could see is that it probably enables this in the REST API then right? Although we generally don't expect people to be using late-binding through the REST API it's still possible. Would we want to explicitly disable this in the non-internal unbind call then?

Reply:
Yeah that's a good idea! I've explicitly disabled unbinding during installation in the non-internal unbind call for the REST API. 

I modified the PR to remove the cancel API I added and I've added the installing statuses to the unbind transition in the statemachine.

What do you think?